Expose elf abi on ppc64 targets

ostylk opened the pull request:
Fixes https://github.com/rust-lang/rust/issues/60617 (after MCP https://github.com/rust-lang/compiler-team/issues/885 is accepted) by exposing the abi information on ppc64 targets.
Conditional compilation can now use `cfg(target_abi = "elfv1")` or `cfg(target_abi = "elfv2")` to determine the abi in use.

Technical details are included in the other PR https://github.com/rust-lang/rust/pull/142598

<!-- homu-ignore:start -->

<details>
<summary> original pull request text </summary>
Attempts to fix https://github.com/rust-lang/rust/issues/85589 (also more information is there).

Basically the problem is that ld.lld assumes that all ppc64 object files with e_flags=0 are object files which use the ELFv2 ABI (this here is the check https://github.com/llvm/llvm-project/blob/main/lld/ELF/Arch/PPC64.cpp#L639).
This pull request sets the correct e_flags to indicate the used ABI so ld.lld errors out when encountering ELFv1 ABI files instead of generating a broken binary.

For example compare code generation for this program (file name ``min.rs``):
```rust
#![feature(no_core, lang_items, repr_simd)]
#![crate_type = "bin"]
#![no_core]
#![no_main]

#[lang = "sized"]
trait Sized {}
#[lang = "copy"]
trait Copy {}
#[lang = "panic_cannot_unwind"]
pub fn panic() -> ! {
    loop {}
}

pub fn my_rad_unmangled_function() {
    loop {}
}

pub fn my_rad_function() {
    loop {}
}

#[no_mangle]
pub fn _start() {
    my_rad_unmangled_function();
    my_rad_function();
}
```
Compile with ``rustc --target=powerpc64-unknown-linux-gnu -C linker=ld.lld -C relocation-model=static min.rs``

Before change:
```
$ llvm-objdump -d min
Disassembly of section .text:
000000001001030c <.text>:
		...
10010334: 7c 08 02 a6  	mflr 0
10010338: f8 21 ff 91  	stdu 1, -112(1)
1001033c: f8 01 00 80  	std 0, 128(1)
10010340: 48 02 00 39  	bl 0x10030378 <_ZN3min25my_rad_unmangled_function17h7471c49af58039f5E>
10010344: 60 00 00 00  	nop
10010348: 48 02 00 49  	bl 0x10030390 <_ZN3min15my_rad_function17h37112b8fd1008c9bE>
1001034c: 60 00 00 00  	nop
		...
```
The branch instructions ``bl 0x10030378`` and ``bl 0x10030390`` are jumping into the ``.opd`` section which is data. That is a broken binary (because fixing those branches is the task of the linker).

After change:
```
error: linking with `ld.lld` failed: exit status: 1
  |
  = note:  "ld.lld" "/tmp/rustcNYKZCS/symbols.o" "<1 object files omitted>" "--as-needed" "-L" "/tmp/rustcNYKZCS/raw-dylibs" "-Bdynamic" "--eh-frame-hdr" "-z" "noexecstack" "-L" "<sysroot>/lib/rustlib/powerpc64-unknown-linux-gnu/lib" "-o" "min" "--gc-sections" "-z" "relro" "-z" "now"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: ld.lld: error: /tmp/rustcNYKZCS/symbols.o: ABI version 1 is not supported
```
Which is correct because ld.lld doesn't support ELFv1 ABI.

</details>

<!-- homu-ignore:end -->

workingjubilee commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -379,6 +379,30 @@ pub(super) fn elf_e_flags(architecture: Architecture, sess: &Session) -> u32 {
+            const EF_PPC64_ABI_ELF_V1: u32 = 1;
+            const EF_PPC64_ABI_ELF_V2: u32 = 2;
+            
+            match sess.target.options.abi.as_ref() {
```
None of the PowerPC targets indicate their ABI version that I'm aware of.

ostylk replied:
That can be changed, there's only a handful of them.

workingjubilee commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -379,6 +379,30 @@ pub(super) fn elf_e_flags(architecture: Architecture, sess: &Session) -> u32 {
+                "elfv2" => EF_PPC64_ABI_ELF_V2,
+                // If options.abi is empty, then try to figure out which ABI to use.
+                // Based loosely on llvm check (check llvm/include/llvm/TargetParser/Triple.h#isPPC64ELFv2ABI and clang/lib/Basic/Targets/PPC.h#PPC64TargetInfo)
+                "" => {
+                    if sess.target.llvm_target.contains("bsd") || sess.target.llvm_target.contains("musl") {
+                        EF_PPC64_ABI_ELF_V2
+                    } else {
+                        EF_PPC64_ABI_ELF_V1
+                    }
+                },
+                // Fall back
+                _ => EF_PPC64_ABI_UNKNOWN,
+            }
```
We should not assume. We should panic, and submit a separate PR proposing to add the elf version to the `abi` field for all PowerPC targets that use an ELF format.

ostylk replied:
Done, but why should the ABI be specified in a seperate PR? Now all predefined ppc64 targets are basically broken?

workingjubilee reviewed (changes requested):
But if we do have anything to do with it, we shouldn't be guessing.

ostylk commented:
Oh thanks, I somehow overlooked llvm_abiname.

I changed it to use llvm_abiname, because it really matters what the llvm codegen does. Also I changed the existing ppc64 targets to set llvm_abiname and abi (in case an app needs that information) accordingly.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15591011458/job/43909931892) [(plain)](https://github.com/rust-lang/rust/commit/f15407ee93260d824b17c8d442a1b6bad3a1e735/checks/43909931892/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

129 LL |     target_abi = "_UNEXPECTED_VALUE",
130    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
131    |
-    = note: expected values for `target_abi` are: ``, `abi64`, `abiv2`, `abiv2hf`, `eabi`, `eabihf`, `fortanix`, `ilp32`, `ilp32e`, `llvm`, `macabi`, `sim`, `softfloat`, `spe`, `uwp`, `vec-extabi`, and `x32`
+    = note: expected values for `target_abi` are: ``, `abi64`, `abiv2`, `abiv2hf`, `eabi`, `eabihf`, `elfv1`, `elfv2`, `fortanix`, `ilp32`, `ilp32e`, `llvm`, `macabi`, `sim`, `softfloat`, `spe`, `uwp`, `vec-extabi`, and `x32`
133    = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration
134 
135 warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args check-cfg/well-known-values.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/check-cfg/well-known-values.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/check-cfg/well-known-values" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--check-cfg=cfg()" "-Zcheck-cfg-all-expected"
stdout: none
--- stderr -------------------------------
warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:29:5
   |
LL |     clippy = "_UNEXPECTED_VALUE",
   |     ^^^^^^----------------------
   |           |
   |           help: remove the value
   |
   = note: no expected value for `clippy`
---

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:33:5
   |
LL |     doc = "_UNEXPECTED_VALUE",
   |     ^^^----------------------
   |        |
   |        help: remove the value
   |
   = note: no expected value for `doc`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration
---

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:37:5
   |
LL |     fmt_debug = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `fmt_debug` are: `full`, `none`, and `shallow`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:39:5
   |
LL |     miri = "_UNEXPECTED_VALUE",
   |     ^^^^----------------------
   |         |
   |         help: remove the value
   |
   = note: no expected value for `miri`
---

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:45:5
   |
LL |     proc_macro = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^----------------------
   |               |
   |               help: remove the value
   |
   = note: no expected value for `proc_macro`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:47:5
   |
LL |     relocation_model = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `relocation_model` are: `dynamic-no-pic`, `pic`, `pie`, `ropi`, `ropi-rwpi`, `rwpi`, and `static`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:49:5
   |
LL |     rustfmt = "_UNEXPECTED_VALUE",
   |     ^^^^^^^----------------------
   |            |
   |            help: remove the value
   |
   = note: no expected value for `rustfmt`
---
   |
LL |     sanitize = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `sanitize` are: `address`, `cfi`, `dataflow`, `hwaddress`, `kcfi`, `kernel-address`, `leak`, `memory`, `memtag`, `safestack`, `shadow-call-stack`, and `thread`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:53:5
   |
LL |     target_abi = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `target_abi` are: ``, `abi64`, `abiv2`, `abiv2hf`, `eabi`, `eabihf`, `elfv1`, `elfv2`, `fortanix`, `ilp32`, `ilp32e`, `llvm`, `macabi`, `sim`, `softfloat`, `spe`, `uwp`, `vec-extabi`, and `x32`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:55:5
   |
LL |     target_arch = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `target_arch` are: `aarch64`, `amdgpu`, `arm`, `arm64ec`, `avr`, `bpf`, `csky`, `hexagon`, `loongarch32`, `loongarch64`, `m68k`, `mips`, `mips32r6`, `mips64`, `mips64r6`, `msp430`, `nvptx64`, `powerpc`, `powerpc64`, `riscv32`, `riscv64`, `s390x`, `sparc`, `sparc64`, `wasm32`, `wasm64`, `x86`, `x86_64`, and `xtensa`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:57:5
   |
---
   |
LL |     target_env = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `target_env` are: ``, `gnu`, `msvc`, `musl`, `newlib`, `nto70`, `nto71`, `nto71_iosock`, `nto80`, `ohos`, `p1`, `p2`, `relibc`, `sgx`, and `uclibc`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:61:5
   |
---

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:65:5
   |
LL |     target_has_atomic = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `target_has_atomic` are: (none), `128`, `16`, `32`, `64`, `8`, and `ptr`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:67:5
   |
LL |     target_has_atomic_equal_alignment = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `target_has_atomic_equal_alignment` are: (none), `128`, `16`, `32`, `64`, `8`, and `ptr`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:69:5
   |
LL |     target_has_atomic_load_store = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `target_has_atomic_load_store` are: (none), `128`, `16`, `32`, `64`, `8`, and `ptr`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:71:5
   |
LL |     target_os = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `target_os` are: `aix`, `amdhsa`, `android`, `cuda`, `cygwin`, `dragonfly`, `emscripten`, `espidf`, `freebsd`, `fuchsia`, `haiku`, `hermit`, `horizon`, `hurd`, `illumos`, `ios`, `l4re`, `linux`, `lynxos178`, `macos`, `netbsd`, `none`, `nto`, `nuttx`, `openbsd`, `psp`, `psx`, `redox`, `rtems`, `solaris`, `solid_asp3`, `teeos`, `trusty`, `tvos`, `uefi`, `unknown`, `visionos`, `vita`, `vxworks`, `wasi`, `watchos`, `windows`, `xous`, and `zkvm`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:73:5
   |
LL |     target_pointer_width = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `target_pointer_width` are: `16`, `32`, and `64`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:75:5
   |
---
   |
LL |     target_vendor = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expected values for `target_vendor` are: `amd`, `apple`, `espressif`, `fortanix`, `ibm`, `kmc`, `mti`, `nintendo`, `nvidia`, `openwrt`, `pc`, `risc0`, `sony`, `sun`, `unikraft`, `unknown`, `uwp`, `win7`, and `wrs`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:79:5
   |
LL |     ub_checks = "_UNEXPECTED_VALUE",
   |     ^^^^^^^^^----------------------
   |              |
   |              help: remove the value
   |
   = note: no expected value for `ub_checks`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:81:5
   |
LL |     unix = "_UNEXPECTED_VALUE",
   |     ^^^^----------------------
   |         |
   |         help: remove the value
   |
   = note: no expected value for `unix`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `_UNEXPECTED_VALUE`
##[warning]  --> /checkout/tests/ui/check-cfg/well-known-values.rs:83:5
   |
LL |     windows = "_UNEXPECTED_VALUE",
   |     ^^^^^^^----------------------
   |            |
   |            help: remove the value
   |
   = note: no expected value for `windows`
---
   |
LL | #[cfg(target_os = "linuz")] // testing that we suggest `linux`
   |       ^^^^^^^^^^^^-------
   |                   |
   |                   help: there is a expected value with a similar name: `"linux"`
   |
   = note: expected values for `target_os` are: `aix`, `amdhsa`, `android`, `cuda`, `cygwin`, `dragonfly`, `emscripten`, `espidf`, `freebsd`, `fuchsia`, `haiku`, `hermit`, `horizon`, `hurd`, `illumos`, `ios`, `l4re`, `linux`, `lynxos178`, `macos`, `netbsd`, `none`, `nto`, `nuttx`, `openbsd`, `psp`, `psx`, `redox`, `rtems`, `solaris`, `solid_asp3`, `teeos`, `trusty`, `tvos`, `uefi`, `unknown`, `visionos`, `vita`, `vxworks`, `wasi`, `watchos`, `windows`, `xous`, and `zkvm`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration

warning: 28 warnings emitted
------------------------------------------

```

</details>

nikic commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -379,6 +379,27 @@ pub(super) fn elf_e_flags(architecture: Architecture, sess: &Session) -> u32 {
+            const EF_PPC64_ABI_ELF_V2: u32 = 2;
+
+            if sess.target.options.llvm_abiname.as_ref().is_empty()
+                && sess.target.options.binary_format.to_object() == BinaryFormat::Elf
```
Presumably this code can't even be reached for non-ELF?

nikic reviewed (commented):
This looks reasonable to me. I also found https://github.com/rust-lang/rust/issues/60617 which asks for exposing the ELF ABI version, so I guess this fixes that issue as well.

Does exposing something new to target_abi require any special sign-off?

workingjubilee reviewed (commented):
My understanding of https://forge.rust-lang.org/compiler/proposals-and-stabilization.html#targets is that it requires an FCP now to add something like a field that will be visible via `target_abi`, since it's user-facing. Or at least MCP.

That is why I suggested splitting up codegen and user-facing parts (I mean people can poke at binaries and so in a strained sense that's "user-facing" but you know what I mean) so that the former is not stuck on process. They'll probably ship in different versions as a result but that seems completely fine.

workingjubilee commented:
I have opened a proposal to clarify we do not need that much process for this sort of additive change. I think that as long as it incorporates the current user-facing elements, it would unfortunately still be stuck on the process of agreeing we want to change the process to require less process. :^)

ostylk commented:
First of all, thank you @workingjubilee for taking care of all the processes required to incorporate this change.

If I understand correctly the "user-facing change" in only the ``abi`` configuration (because those may be used by conditional compilation ``#[cfg..]``). So a seperate PR without the ``abi`` changes but keeping the ``llvm_abiname`` changes wouldn't be blocked, right?
If that's the case, then I can do that (because that already fixes https://github.com/rust-lang/rust/issues/85589).
Then after this PR is unblocked and merged, that fixes https://github.com/rust-lang/rust/issues/60617 .

ostylk committed:
expose abi information on ppc64 targets

ostylk commented:
I split this PR in two.

https://github.com/rust-lang/rust/pull/142598 contains the pure code generation changes.
This pull request contains the user-facing changes.

When the other PR is merged, I'll rebase this to only include the last commit.

programmerjake commented:
Going through the list of 64-bit powerpc targets, the only missing one is `powerpc64-ibm-aix` but that's fine since it's not an ELF target.

Afaict 32-bit targets (don't have ELF ABI v1/v2 stuff)[https://github.com/bminor/binutils-gdb/blob/564624a452d0f6fd6fb87c40ed1d1ef866fd493e/gas/config/tc-ppc.c#L2598-L2606].

bors commented:
:pushpin: Commit cfa6731c495192a490492e8aae91425a9825223b has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee cfa6731c495192a490492e8aae91425a9825223b -->
<!-- homu: {"type":"Approved","sha":"cfa6731c495192a490492e8aae91425a9825223b","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
